### ISSUE TITLE
Missing and unexpected diagnostics for a delegate assignment

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#29844]
```

        [Fact]
        public void NonNullTypes_Delegate()
        {
            var source = @"
using System.Runtime.CompilerServices;

[NonNullTypes(true)]
public delegate string[] MyDelegate(string[] x);

[NonNullTypes(false)]
public delegate string[] MyFalseDelegate(string[] x);

public delegate string[]? MyNullableDelegate(string[]? x);

class C
{
    void M()
    {
        MyDelegate x1 = Method;
        MyDelegate x2 = FalseMethod;
        MyDelegate x4 = NullableReturnMethod; // warn 1
        MyDelegate x5 = NullableParameterMethod; // warn 2
        MyFalseDelegate y1 = Method;
        MyFalseDelegate y2 = FalseMethod;
        MyFalseDelegate y4 = NullableReturnMethod;
        MyFalseDelegate y5 = NullableParameterMethod;
        MyNullableDelegate z1 = Method; // warn 3
        MyNullableDelegate z2 = FalseMethod;
        MyNullableDelegate z4 = NullableReturnMethod; // warn 4
        MyNullableDelegate z5 = NullableParameterMethod; // warn 5
     }

    [NonNullTypes(true)]
    public string[] Method(string[] x) => throw null;

    [NonNullTypes(false)]
    public string[] FalseMethod(string[] x) => throw null;

    public string[]? NullableReturnMethod(string[] x) => throw null;
    public string[] NullableParameterMethod(string[]? x) => throw null;
}
";
            var compilation = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });

            compilation.VerifyTypes();
            compilation.VerifyDiagnostics(
                // (18,25): warning CS8621: Nullability of reference types in return type of 'string[]? C.NullableReturnMethod(string[] x)' doesn't match the target delegate 'MyDelegate'.
                //         MyDelegate x4 = NullableReturnMethod; // warn 1
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInReturnTypeOfTargetDelegate, "NullableReturnMethod").WithArguments("string[]? C.NullableReturnMethod(string[] x)", "MyDelegate").WithLocation(18, 25),
                // (24,33): warning CS8622: Nullability of reference types in type of parameter 'x' of 'string[] C.Method(string[] x)' doesn't match the target delegate 'MyNullableDelegate'.
                //         MyNullableDelegate z1 = Method; // warn 3
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "Method").WithArguments("x", "string[] C.Method(string[] x)", "MyNullableDelegate").WithLocation(24, 33),
                // (26,33): warning CS8622: Nullability of reference types in type of parameter 'x' of 'string[]? C.NullableReturnMethod(string[] x)' doesn't match the target delegate 'MyNullableDelegate'.
                //         MyNullableDelegate z4 = NullableReturnMethod; // warn 4
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "NullableReturnMethod").WithArguments("x", "string[]? C.NullableReturnMethod(string[] x)", "MyNullableDelegate").WithLocation(26, 33)
                );

            // PROTOTYPE(NullableReferenceTypes): Missing warnings 2 and 5
        }
```

**Comment by AlekseyTs** [comment]
```
        [Fact]
        public void IdentityConversion_DelegateParameter_02()
        {
            var source =
@"delegate T D<T>();
class A<T>
{
    internal T M() => throw new System.NotImplementedException();
}
class B
{
    static A<T> F<T>(T t) => throw null;
    static void G(object? o)
    {
        var x = F(o);
        D<object?> d = x.M;
        D<object> e = x.M;
        if (o == null) return;
        var y = F(o);
        d = y.M;
        e = y.M;
        d = (D<object?>)y.M;
        e = (D<object>)y.M;
    }
}";
            var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
            // PROTOTYPE(NullableReferenceTypes): Should report WRN_NullabilityMismatchInReturnTypeOfTargetDelegate for `e = x.M`.
            comp.VerifyDiagnostics();
        }
```

**Comment by AlekseyTs** [comment]
```
        [Fact]
        public void IdentityConversion_DelegateOutParameter()
        {
            var source =
@"delegate void D<T>(out T t);
interface I<T> { }
interface IIn<in T> { }
interface IOut<out T> { }
class C
{
    static void F<T>(out T t) { t = default; }
    static void G()
    {
        D<object> a = F<object?>;
        D<object?> b = F<object>;
        D<I<object>> c = F<I<object?>>;
        D<I<object?>> d = F<I<object>>;
        D<IIn<object>> e = F<IIn<object?>>;
        D<IIn<object?>> f = F<IIn<object>>;
        D<IOut<object>> g = F<IOut<object?>>;
        D<IOut<object?>> h = F<IOut<object>>;
    }
}";
            var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
            // PROTOTYPE(NullableReferenceTypes): Should not warn for `b`, `e`, `h`.
            comp.VerifyDiagnostics(
                // (7,37): warning CS8625: Cannot convert null literal to non-nullable reference or unconstrained type parameter.
                //     static void F<T>(out T t) { t = default; }
                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "default").WithLocation(7, 37),
                // (10,23): warning CS8622: Nullability of reference types in type of parameter 't' of 'void C.F<object?>(out object? t)' doesn't match the target delegate 'D<object>'.
                //         D<object> a = F<object?>;
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<object?>").WithArguments("t", "void C.F<object?>(out object? t)", "D<object>").WithLocation(10, 23),
                // (11,24): warning CS8622: Nullability of reference types in type of parameter 't' of 'void C.F<object>(out object t)' doesn't match the target delegate 'D<object?>'.
                //         D<object?> b = F<object>;
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<object>").WithArguments("t", "void C.F<object>(out object t)", "D<object?>").WithLocation(11, 24),
                // (12,26): warning CS8622: Nullability of reference types in type of parameter 't' of 'void C.F<I<object?>>(out I<object?> t)' doesn't match the target delegate 'D<I<object>>'.
                //         D<I<object>> c = F<I<object?>>;
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<I<object?>>").WithArguments("t", "void C.F<I<object?>>(out I<object?> t)", "D<I<object>>").WithLocation(12, 26),
                // (13,27): warning CS8622: Nullability of reference types in type of parameter 't' of 'void C.F<I<object>>(out I<object> t)' doesn't match the target delegate 'D<I<object?>>'.
                //         D<I<object?>> d = F<I<object>>;
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<I<object>>").WithArguments("t", "void C.F<I<object>>(out I<object> t)", "D<I<object?>>").WithLocation(13, 27),
                // (14,28): warning CS8622: Nullability of reference types in type of parameter 't' of 'void C.F<IIn<object?>>(out IIn<object?> t)' doesn't match the target delegate 'D<IIn<object>>'.
                //         D<IIn<object>> e = F<IIn<object?>>;
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<IIn<object?>>").WithArguments("t", "void C.F<IIn<object?>>(out IIn<object?> t)", "D<IIn<object>>").WithLocation(14, 28),
                // (15,29): warning CS8622: Nullability of reference types in type of parameter 't' of 'void C.F<IIn<object>>(out IIn<object> t)' doesn't match the target delegate 'D<IIn<object?>>'.
                //         D<IIn<object?>> f = F<IIn<object>>;
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<IIn<object>>").WithArguments("t", "void C.F<IIn<object>>(out IIn<object> t)", "D<IIn<object?>>").WithLocation(15, 29),
                // (16,29): warning CS8622: Nullability of reference types in type of parameter 't' of 'void C.F<IOut<object?>>(out IOut<object?> t)' doesn't match the target delegate 'D<IOut<object>>'.
                //         D<IOut<object>> g = F<IOut<object?>>;
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<IOut<object?>>").WithArguments("t", "void C.F<IOut<object?>>(out IOut<object?> t)", "D<IOut<object>>").WithLocation(16, 29),
                // (17,30): warning CS8622: Nullability of reference types in type of parameter 't' of 'void C.F<IOut<object>>(out IOut<object> t)' doesn't match the target delegate 'D<IOut<object?>>'.
                //         D<IOut<object?>> h = F<IOut<object>>;
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<IOut<object>>").WithArguments("t", "void C.F<IOut<object>>(out IOut<object> t)", "D<IOut<object?>>").WithLocation(17, 30));
        }
```

**Comment by cston** [comment]
All issues issues have been addressed.
